stl/src/cthread.cpp

```diff
@@ -103,9 +105,45 @@ _CRTIMP2_PURE _Thrd_id_t __cdecl _Thrd_id() noexcept { // return unique id for c
 }
 
 _CRTIMP2_PURE unsigned int __cdecl _Thrd_hardware_concurrency() noexcept { // return number of processors
-    SYSTEM_INFO info;
-    GetNativeSystemInfo(&info);
-    return info.dwNumberOfProcessors;
+    // Most devices have only one processor group and thus have the same buffer_size
+#if defined(_M_X64) || defined(_M_ARM64) // 16 bytes per group
+    constexpr int stack_buffer_size = 48;
+#elif defined(_M_IX86) || defined(_M_ARM) // 12 bytes per group
+    constexpr int stack_buffer_size = 44;

```

Comment:
Would it be better to replace these with `offsetof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX, Processor) + sizeof(PROCESSOR_RELATIONSHIP)` to avoid hard coding?

Language-lawyer-ism: The initial buffer size is smaller than `sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX)`, which is quite scary.

Reply:
I think that's more complicated to reason about. This is a squirrely API, but because it's given the size of the buffer, it will accurately report whether it's able to return the requested information (and if so, how much it wrote) or whether it had insufficient space (and if so, how much it would need). As a result, the stack buffer size is not correctness-critical. If it's too large, the space will just be wasted but the function will succeed on the first try. If it's too small, the first try will fail, and we'll dynamically allocate.

I manually verified on x64 and x86 that these values are accurate for the normal one-socket one-group case.